### PR TITLE
FED-2855 Fix Analyzer Plugin error when typing keys

### PR DESCRIPTION
**Pull request by greglittlefield-wf**
## Motivation
When a user's in the middle of typing an identifier to be used as a key, the plugin currently throws an error due to a bad cast:

```dart
example(someIdentifier) {
  (Dom.div()
    ..key = someIden/*user's cursor here, typing*/
  )();
}
```

```
type 'InvalidTypeImpl' is not a subtype of type 'InterfaceType' in type cast

#0      BadKeyDiagnostic.processToStringedExpressionInKey (package:over_react_analyzer_plugin/src/diagnostic/bad_key.dart:155:48)
#1      BadKeyDiagnostic.computeErrorsForUsage (package:over_react_analyzer_plugin/src/diagnostic/bad_key.dart:119:7)
#2      ComponentUsageDiagnosticContributor.computeErrorsForUsages 
...
```

## Changes
- Handle `InvalidType` case, which as of analyzer 5.13.0 is used for undefined identifiers
   - Supporting both analyzer <5.13.0 and >=5.13.0 in the same code was difficult, so now that we no longer support Dart 2.18 (which can't resolve to analyzer >=5.13.0), I've updated the lower bound
- Replace unsafe `as InterfaceType` cast with safe `is InterfaceType` check, clean up surrounding logic
- Add regression test for non-empty undefined identifier

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        - Regression test fails in the commit it was added in and passes on latest commit, which contains the fix
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Frontend Frameworks Design member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: tools/analyzer_plugin/lib/src/diagnostic/bad_key.dart [hunk 1]
```diff
@@ -162,7 +165,7 @@ class BadKeyDiagnostic extends ComponentUsageDiagnosticContributor {
 
       if (type.isDartCoreInt || type.isDartCoreDouble || type.isDartCoreString || type.isDartCoreSymbol) {
         // Ignore core types that have good `Object.toString` implementations values.
-      } else if (type.isDartCoreObject || type.isDynamic) {
+      } else if (type.isDartCoreObject || type is DynamicType) {
```
Comment (greglittlefield-wf) [inline tools/analyzer_plugin/lib/src/diagnostic/bad_key.dart:168]: This was a deprecation fix

File: tools/analyzer_plugin/lib/src/diagnostic/bad_key.dart [hunk 2]
```diff
@@ -139,21 +139,24 @@ class BadKeyDiagnostic extends ComponentUsageDiagnosticContributor {
...
     if (topLevelKeyType == null) return;
+    // The type is bad because of errors, such as when the user's in the middle of typing an identifier,
+    // or referencing an identifier that doesn't exist yet.
+    if (topLevelKeyType is InvalidType) return;
 
     final typedProvider = result.typeProvider;
 
     // Special-case for Iterables and Maps: check if their (keys and) values are allowed types.
     final isIterableOrMap =
+        topLevelKeyType is! DynamicType &&
```
Comment (greglittlefield-wf) [inline tools/analyzer_plugin/lib/src/diagnostic/bad_key.dart:150]: This boolean was previously true if the type was dynamic

**Review by sydneyjodon-wk — APPROVED** [review]
+10 🎉 

- [x] verified [new test fails](https://github.com/Workiva/over_react/actions/runs/9650502131/job/26616251373?pr=929) with the expected error and these changes fix that test

@Workiva/release-management-p

**Comment by greglittlefield-wf** [comment]
Created a new PR here: https://github.com/Workiva/over_react/pull/930
